### PR TITLE
Cache lookups

### PR DESCRIPTION
**Pull request by KubaJagiello**
Adding cache that solves the #22 issue.

### CODE REVIEW

File: backend/web/web_backend/src/cache.rs [hunk 1]
```diff
@@ -1,28 +1,197 @@
...
+
+    for i in 0..50 {
+        cache.store("abc", i).await;
+    }
+
+    assert_eq!(cache.max_size.unwrap().read().await.filo.len(), 5);
+}
+*/
+
+// TODO: Setup benchmark
```
Comment (frepet) [inline backend/web/web_backend/src/cache.rs]: This should be implemented before this is pulled to master.

File: backend/web/web_backend/src/user_provider.rs [hunk 2]
```diff
@@ -85,6 +85,9 @@ impl UserProvider {
     /// TODO: Make UserProvider actually use a configured user provider to fetch data
     /// TODO: Cache lookups to user provider
     async fn get_user_from_provider(&self, user_id: &str) -> anyhow::Result<RemoteUser> {
+        // Lookup cache
+        //
+
```
Comment (frepet) [inline backend/web/web_backend/src/user_provider.rs]: What is this?

File: backend/web/web_backend/src/cache.rs [hunk 3]
```diff
@@ -1,28 +1,197 @@
...
+        CacheBuilder::default()
+    }
+
+    pub async fn lookup(&self, key: &str) -> Option<Arc<V>> {
+        // If the timeout is set
+        if let Some(timeout) = &self.timeout {
+            // If the entry is too old
+            if !timeout.write().await.is_valid(String::from(key)) {
+                // TODO: if this is invalid, MaxSize will contain one unnecessary
+                // entry
```
Comment (frepet) [inline backend/web/web_backend/src/cache.rs]: 144 makes lookup grab a write lock for every lookup regardless if it is needed or not.
Is the TODO a problem? If it is not, remove the TODO, if it is, it should be resolved before the pull.

File: backend/web/web_backend/src/cache.rs [hunk 4]
```diff
@@ -1,28 +1,197 @@
...
+// Emluate queue where we can save the Keys from cache
+// As long as the queue is not filled, we save all the data
+
+// When the queue hits maximum, we will start replacing the
+// oldest entries.
+impl MaxSize {
+    fn new(max_size: usize) -> Self {
+        MaxSize {
+            max_size,
+            filo: Vec::new(), // first-in-last-out
```
Comment (frepet) [inline backend/web/web_backend/src/cache.rs:53]: Is the concept of the ring buffer like queue over-engineered?
Would a deque do the work in an easier to read fashion while still maintaining high performance?

**Comment by frepet** [comment]
Did you do `cargo fmt` on the files? Looks like you forgot it.

**Review by KlasafGeijerstam — APPROVED** [review]

**Pull request merged**